### PR TITLE
Pass size of source while calling put for storage bucket from kando

### PR DESCRIPTION
**Pull request by viveksinghggits**
## Change Overview

We were passing the object size `0` when we called `.put` on
the bucket of object storage and because of that whenever `stow`
tried to check if source size is more than the allowed size we didn't 
get expected result.
This PR tries to fix that by passing the correct object size while calling `put`.

Depends on: https://github.com/kanisterio/kanister/pull/809

## Pull request type

Please check the type of change your PR introduces:

- [x] :bug: Bugfix

## Issues

- #XXX

## Test Plan

```
» cd pkg/location
» go test                                        
OK: 4 passed, 2 skipped
PASS
ok  	github.com/kanisterio/kanister/pkg/location	12.187s

```

I tried to upload the file of size `1000M`, if the size is greater that `256M` multipart upload for stow would be called.
```
» ls -lh
-rw-rw-r-- 1 vivek vivek 1000M Nov  5 18:29 1000M.db.gz

» ./kando location push --profile '{"Location":{"type":"azure","bucket":"vivek-test","endpoint":"","prefix":"k10/480b12f5-9107-41cb-9f18-ad0b57165d08/migration","region":"Central US"},"Credential":{"Type":"keyPair","KeyPair":{"ID":"test","Secret":"test"},"Secret":null},"SkipSSLVerify":false}'  ./1000M.db.gz --path testsize/1/1000
```

- [x] :muscle: Manual
- [ ] :zap: Unit test
- [ ] :green_heart: E2E

### CODE REVIEW

**Comment by viveksinghggits** [comment]
Please ignore the `go.mod`, `go.sum` changes as of now.

File: pkg/location/location.go [hunk 1]
```diff
@@ -97,7 +98,12 @@ func writeData(ctx context.Context, pType objectstore.ProviderType, profile para
 	if err != nil {
 		return err
 	}
-	if err := bucket.Put(ctx, path, in, 0, nil); err != nil {
+	b, err := ioutil.ReadAll(in)
```
Comment (vkamra) [inline pkg/location/location.go:101]: Won't we end up reading the data twice?

Comment (viveksinghggits) [reply pkg/location/location.go:101]: yeah, we discussed this in yesterday's call. We unknowingly introduced another issue here. After reading the `in` is going to be `nil`. I am working on this.

**Comment by viveksinghggits** [comment]
This requires the version of `kastenhq/stow` that has multipart support (https://github.com/graymeta/stow/pull/162). The tag that we were using in kanister doesn't have that code.

**Comment by PrasadG193** [comment]
Unit tests run after updating `stow` dep to `v0.2.6-kasten`
```
$ go test -v . -check.vv -check.suitep=4 -timeout=50m
=== RUN   Test
START: location_test.go:60: LocationSuite.SetUpSuite
SKIP: location_test.go:60: LocationSuite.SetUpSuite (Test  requires the environemnt variable 'GOOGLE_APPLICATION_CREDENTIALS')

START: location_test.go:138: LocationSuite.TestAzMultipartUpload
SKIP: location_test.go:138: LocationSuite.TestAzMultipartUpload

START: location_test.go:175: LocationSuite.TestReaderSize
SKIP: location_test.go:175: LocationSuite.TestReaderSize

START: location_test.go:127: LocationSuite.TestWriteAndReadData
SKIP: location_test.go:127: LocationSuite.TestWriteAndReadData

START: location_test.go:60: LocationSuite.SetUpSuite
PASS: location_test.go:60: LocationSuite.SetUpSuite     7.730s

START: location_test.go:138: LocationSuite.TestAzMultipartUpload
START: location_test.go:108: LocationSuite.TearDownTest
PASS: location_test.go:108: LocationSuite.TearDownTest  4.769s

SKIP: location_test.go:138: LocationSuite.TestAzMultipartUpload (Not applicable for location type S3)

START: location_test.go:175: LocationSuite.TestReaderSize
START: location_test.go:108: LocationSuite.TearDownTest
PASS: location_test.go:108: LocationSuite.TearDownTest  0.628s

PASS: location_test.go:175: LocationSuite.TestReaderSize        0.000s

START: location_test.go:127: LocationSuite.TestWriteAndReadData
START: location_test.go:108: LocationSuite.TearDownTest
PASS: location_test.go:108: LocationSuite.TearDownTest  0.593s

PASS: location_test.go:127: LocationSuite.TestWriteAndReadData  16.202s

START: location_test.go:60: LocationSuite.SetUpSuite
PASS: location_test.go:60: LocationSuite.SetUpSuite     0.229s

START: location_test.go:138: LocationSuite.TestAzMultipartUpload
START: location_test.go:108: LocationSuite.TearDownTest
Cannot cleanup test directory: 2020-11-11T05:37:23.474016906Z/testlocation.txt
PASS: location_test.go:108: LocationSuite.TearDownTest  0.093s

PASS: location_test.go:138: LocationSuite.TestAzMultipartUpload 704.459s

START: location_test.go:175: LocationSuite.TestReaderSize
START: location_test.go:108: LocationSuite.TearDownTest
Cannot cleanup test directory: 2020-11-11T05:37:23.474016906Z/testlocation.txt
Cannot cleanup test directory: 2020-11-11T05:37:23.474016906Z/testchunk.txt
PASS: location_test.go:108: LocationSuite.TearDownTest  0.082s

PASS: location_test.go:175: LocationSuite.TestReaderSize        0.000s

START: location_test.go:127: LocationSuite.TestWriteAndReadData
START: location_test.go:108: LocationSuite.TearDownTest
Cannot cleanup test directory: 2020-11-11T05:37:23.474016906Z/testchunk.txt
PASS: location_test.go:108: LocationSuite.TearDownTest  0.056s

PASS: location_test.go:127: LocationSuite.TestWriteAndReadData  3.959s

OK: 5 passed, 4 skipped
--- PASS: Test (708.88s)
PASS
ok      github.com/kanisterio/kanister/pkg/location     708.973s
```

**Comment by PrasadG193** [comment]
1. Test AZ upload with empty files

```
$ ll -l dump 
-rw-rw-r-- 1 prasad prasad 0 Nov 11 10:25 dump

# ./dist/kando_linux_amd64/kando location push --profile '{"Location":{"type":"azure","bucket":"xxxxxxxxx","endpoint":"","prefix":"k10/test-migration","region":"westindia"},"Credential":{"Type":"keyPair","KeyPair":{"ID":"xxxxx","Secret":"xxxxxxxxxxxxxxxx"},"Secret":null},"SkipSSLVerify":false}'   dump --path testsize/0

# az storage blob list --container-name prasad-test --account-name xxxxxxxxxxx --account-key xxxxxxxxxxxxx 

  {                                                                                                                                                                                                                
    "content": null,                                                                                                                                                                                               
    "deleted": false,                                                                                                                                                                                              
    "metadata": null,                                                                                    
    "name": "k10/test-migration/testsize/0",                                                             
    "properties": {                                                                                                                                                                                                
      "appendBlobCommittedBlockCount": null,                                                             
      "blobTier": "Hot",                      
      "blobTierChangeTime": null,                                                                        
      "blobTierInferred": true,                                                                          
      "blobType": "BlockBlob",                                                                                                                                                                                     
      "contentLength": 0,                                                                                                                                                                                          
      "contentRange": null,      
      "contentSettings": {                                                                                                                                                                                         
        "cacheControl": null,
        "contentDisposition": null,
        "contentEncoding": null,                                                                         
        "contentLanguage": null,
        "contentMd5": "1B2M2Y8AsgTpgAmY7PhCfg==",
        "contentType": "application/octet-stream"
      },
.
.
.

```


2. Test Azure upload with 2G file

```
$ ll -l dump 
-rw-rw-r-- 1 prasad prasad 2147483648 Nov 10 18:29 dump

$ ./dist/kando_linux_amd64/kando location push --profile '{"Location":{"type":"azure","bucket":"xxxxxxxxx","endpoint":"","prefix":"k10/480b12f5-9107-41cb-9f18-ad0b57165d08/migration","region":"westindia"},"Credential":{"Type":"keyPair","KeyPair":{"ID":"xxxxx","Secret":"xxxxxxxxxxxxxxxx"},"Secret":null},"SkipSSLVerify":false}'   dump --path estsize/1

$ az storage blob list --container-name prasad-test --account-name xxxxxxxxxxx --account-key xxxxxxxxxxxxx 

  {                                                                                                                                                   
    "content": null,                                                                                                                                  
    "deleted": false,                                                                                                                                 
    "metadata": null,                                                      
    "name": "k10/480b12f5-9107-41cb-9f18-ad0b57165d08/migration/testsize/1",                  
    "properties": {                                                        
      "appendBlobCommittedBlockCount": null,                                                                                                          
      "blobTier": "Hot",                                                                                                                              
      "blobTierChangeTime": null,
      "blobTierInferred": true,                                                                                                                       
      "blobType": "BlockBlob",                                             
      "contentLength": 2147483648,                                         
      "contentRange": null,                                                
      "contentSettings": {
        "cacheControl": null,
        "contentDisposition": null,
        "contentEncoding": null,
        "contentLanguage": null,
        "contentMd5": null,
        "contentType": "application/octet-stream"
      },
```


3. Test S3 upload for 0 filesize

```
$ ls -l dump 
-rw-rw-r-- 1 prasad prasad 0 Nov 11 09:55 dump


./dist/kando_linux_amd64/kando location push --profile '{"Location":{"type":"s3Compliant","bucket":"xxxxx","endpoint":"","prefix":"k10/test-migration","region":"ap-south-1"},"Credential":{"Type":"keyPair","KeyPair":{"ID":"xxxxxx","Secret":"xxxxxxxxx+9z0vyf"},"Secret":null},"SkipSSLVerify":false}' dump --path testsize/0

$ aws s3 ls s3://xxxxxx/k10/test-migration/ --recursive
2020-11-11 10:01:52          0 k10/test-migration/testsize/0
```


4. Test S3 upload for 10M file

```
$ ls -l dump 
-rw-rw-r-- 1 prasad prasad 10485760 Nov 11 10:04 dump

$ ./dist/kando_linux_amd64/kando location push --profile '{"Location":{"type":"s3Compliant","bucket":"xxxxxx","endpoint":"","prefix":"k10/test-migration","region":"ap-south-1"},"Credential":{"Type":"keyPair","KeyPair":{"ID":"xxxx","Secret":"xxxxxxxx"},"Secret":null},"SkipSSLVerify":false}' dump --path testsize/100


$ aws s3 ls s3://xxxxxxxx/k10/test-migration/ --recursive
2020-11-11 10:01:52          0 k10/test-migration/testsize/0
2020-11-11 10:04:32   10485760 k10/test-migration/testsize/100

File: pkg/location/location_test.go [hunk 2]
```diff
@@ -107,18 +112,95 @@ func (s *LocationSuite) TearDownTest(c *C) {
...
+	}
+	// Create test file
+	f, err := os.Create(s.testMultipartPath)
+	c.Check(err, IsNil)
+	defer f.Close()
+
+	ctx := context.Background()
+	for _, fileSize := range []int64{
+		0,                 // empty file
+		100 * 1024 * 1024, // 100M ie < buffSize
```
Comment (tdmanv) [inline pkg/location/location_test.go:157]: Please test the boundary cases - 256MiB -1, 256MiB, 256MiB+1

File: pkg/location/location_test.go [hunk 3]
```diff
@@ -107,18 +112,95 @@ func (s *LocationSuite) TearDownTest(c *C) {
...
+		300 * 1024 * 1024, // 300M ie > buffSize
+	} {
+		// Create dump file
+		err := os.Truncate(s.testMultipartPath, fileSize)
+		c.Assert(err, IsNil)
+		testcontent, err := ioutil.ReadFile(s.testMultipartPath)
```
Comment (tdmanv) [inline pkg/location/location_test.go:163]: `os.Open()` will give you a reader - you don't need to get the content in-memory then create a bytes.Buffer.

**Review by tdmanv — APPROVED** [review]

File: pkg/location/location_test.go (hunk 2 above)
Comment (PrasadG193) [reply pkg/location/location_test.go:157]: Sure

File: pkg/location/location_test.go (hunk 3 above)
Comment (PrasadG193) [reply pkg/location/location_test.go]: We are re-using file pointer created at https://github.com/kanisterio/kanister/pull/809/files/147b928221ac56008f56f83ce76f2e525c764eee#diff-c1b47cb14ad134412666df8eff6311f52edc385a3e6ffa6a15564f24ec9b297bR150
